src/Database/LSMTree/Internal/Run/Index/Compact.hs

```diff
@@ -68,80 +77,286 @@ import           Database.LSMTree.Internal.Serialise
+  @O(nk)@ bits. Say we have @O(n)@ pages and we store only the @m@ most
+  significant bits, then the memory size of @Index3@ is @O(n m
+  + 2 n + E[collision n m] * k)@ bits, where the last summand is the expected
+  number of collisions for @n@ independently generated hashes of bit-size @m@.
+  So, @Index3@ is not strictly an improvement over @Index2@ if we look at memory
+  complexity alone. However, /in practice/ @Index3@ is an improvement over
+  @Index2@ if we can pick an @m@ that is (i) much smaller than @k@ and (ii)
+  keeps @E[collision n m]@ small (if not close to @0@).
+
+  For example, storing the first @48@ bits of @2.5@ million SHA256 hashes
+  reduces memory size from @n * 256@ bits to @n * 48@ bits, because the expected
+  number of collisions is smaller than @1@.
+
+  TODO: the expected number of collisions is (n^2 - n ) / (2 * m) for hashing
+  @n@ distinct values into @m@ distinct buckets. Is the following reasoning
+  correct?
+
+  \[
+  \begin{align*}
+     &~ O \left( \frac{n^2 - n}{2m}                                                 \right) \\
+    =&~ O \left( \left( n^n - n \right) *   \left( \frac{1}{2m}             \right) \right) \\
+    =&~ O        \left( n^n - n \right) * O \left( \frac{1}{2m}                     \right) \\
+    =&~ O        \left( n^2 - n \right) * O \left( \left( 2m \right)^{ -1 }         \right) \\
+    =&~ O        \left( n^2 - n \right) * O \left( 2^{ -1} * m^{ -1}                \right) \\
+    =&~ O        \left( n^2 - n \right) * O \left( m^{ -1}                          \right) \\
+    =&~ O        \left( n^2     \right) * O \left( m^{ -1}                          \right) \\
+    =&~ O \left(        n^2             *          m^{ -1}                          \right) \\
+    =&~ O \left( \frac{n^2}{m}                                                      \right) \\
+  \end{align*}
+  \]
+
+  ==== Range-finder
+
+  TODO: explain range-finder bits
+
+  === A semi-formal description of the compact index #rep-descr#
+
+  * \(n\) is the number of pages
+  * \(ps = \{p_i \mid 0 \leq i < n \}\) is a sorted set of pages
+  * \(p^{min}_i\) is the full minimum key on a page \(p_i \in ps\).
+  * \(p^{max}_i\) is the full maximum key on a page \(p_i \in ps\).
+  * \(m\) is the range-finder bit-precision
+  * \(\texttt{topBits16}(m, k)\) extracts the @m@ most significant bits from
+    @k@. We call these @m@ bits the range-finder bits.
+  * \(\texttt{sliceBits32}(m, k)\) extracts the @32@ most significant bits from @k@
+    /after/ @m@. We call these @32@ bits the primary bits.
+  * \(i \in \left[0, n \right)\), unless stated otherwise
+  * \(j \in \left[0, 2^m\right)\), unless stated otherwise
+  * Pages must be partitioned: \(\forall p \in ps. \texttt{topBits16}(m, p^{min}_x) ~\texttt{==}~ \texttt{topits16}(m,p^{max}_x) \)
+
+  \[
+  \begin{align*}
+    RF     :&~ \texttt{Array Word16 PageNo} \\
+    RF[j]   =&~ \underset{x \text{ s.t. } 0 \leq x < n}{argmin} x \leq \texttt{topBits16}(m, p^{min}_x) \\
```

Comment:
argmin?

Reply:
https://en.wikipedia.org/wiki/Arg_max 

I have to fix up this equation. I think I meant to come back to it later while writing but eventually forgot to.

The idea is that $RF[j]$ should point to the first page that has a minimum key with $j$ as its range-finder bits. If no such page exists, then $RF[j]$ should instead point to the first page that has a minimum key with $j+1$ as its range-finder bits. If no such page exists, then similar for $j+2$, etc.

So, I arrive at this equation: 

```math
\begin{align*}
    RF     :&~ \texttt{Array Word16 PageNo} \\
    RF[j]   =&~ \min \{ i \mid j \leq \texttt{topBits16}(m, p^{min}_i) \}  \\
    RF[2^m] =&~ n \\
\end{align*}
```

Example:

```math
\begin{align*}
  \texttt{topBits16}(2, p^{min}_0) =&~ 00 \\ 
  \texttt{topBits16}(2, p^{min}_1) =&~ 00 \\ 
  \texttt{topBits16}(2, p^{min}_2) =&~ 11 \\
  \\
  RF[00] =&~ 0 \\
  RF[01] =&~ 2 \\
  RF[10] =&~ 2 \\
  RF[11]  =&~ 2 \\
  RF[100] =&~ 3 \\
\end{align*} 
```

So the intervals $\left[RF[0], RF[1] \right)$ and $\left[RF[3], RF[4] \right)$ are now non-empty and the other intervals (like $\left[RF[1], RF[2] \right)$ ) are empty, as they should be


TIL: I can write mathematical expressions in GitHub text fields using LateX formatting ([link to the docs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/writing-mathematical-expressions#about-writing-mathematical-expressions)). It uses `MathJax` to render the math, which is the same tool that `haddock` uses!